File: scrape/border-crossings.js

var domain = 'https://en.wikipedia.org';


var scraperjs = require('scraperjs');

var scraper = scraperjs.StaticScraper;

console.log('Scraping wikipedia border crossings');
console.log('This will take a while...', "\n");


// TODO: use $.Deferred
// http://taoofcode.net/promise-anti-patterns/

function ms_delay () {
    return 500;
}
var count = 0;
function delay_msg () {
    count++;
    console.log(count);
}

scraper.create('https://en.wikipedia.org/wiki/Category:International_border_crossings')
    .scrape(function($) {
        return $("#mw-subcategories ul li a").map(function() {
            return {
                country: $(this).text().replace('Border crossings of ', ''),
                link: domain + $(this).attr('href')
            };
        }).get();
    }).then(function(countries) {
        // console.log(countries);
        for (var i = 0, s = countries.length; i < s; i++) {
            // scrape the country-level borders
            scraper.create(countries[i].link)
                .delay(ms_delay(), delay_msg)
                .scrape(function($) {
                    return $("#mw-subcategories ul li a").map(function() {
                        return domain + $(this).attr('href');
                    }).get();
                })
                .then(function(border_pages) {
                    // console.log(border_pages);
                    // not really interested in this level
                    for (var j = 0, t = border_pages.length; j < t; j++) {
                        // scrape the country-border
                        scraper.create(border_pages[j])
                            .delay(ms_delay(), delay_msg)
                            .scrape(function($) {
                                return $("#mw-pages ul li a").map(function() {
                                    return domain + $(this).attr('href');
                                }).get();
                            })
                            .then(function(border_crossings) {
                                // this is the level with the border crossings at a specific border
                                // console.log(border_crossings);
                                for (var k = 0, u = border_crossings.length; k < u; k++) {
                                    // scrape the specific border crossing
                                    scraper.create(border_crossings[k])
                                        .scrape(function($) {
                                            var geo_page = $('#coordinates a.external').attr('href');
                                            if (geo_page) {
                                                // invalid URI?
                                                if (geo_page.indexOf('//') === 0) {
                                                    geo_page = 'http:' + geo_page;
                                                }
                                                // go one step further to fetch the un-formatted coords
                                                var b = scraper.create(geo_page)
                                                    //.delay(ms_delay(), delay_msg)
                                                    .scrape(function($) {
                                                        return $("#mw-pages ul li a").map(function() {
                                                            return domain + $(this).attr('href');
                                                        }).get();
                                                    })
                                                    .then(function(border_crossings) {
                                                        return {
                                                            latitude:  $('.geo').first().find('.latitude').text(),
                                                            longitude: $('.geo').first().find('.longitude').text()
                                                        };
                                                    });

                                                // capture the name of the border crossing
                                                b.name = $('#firstHeading').text();

                                                // border crossing
                                                console.log(b);
                                                return b;
                                            }
                                    }).then(function(b) {

                                        // console.log(b);

                                        // save the border crossings...



                                    });
                                }// for
                        });
                    }// for
                });
        }// for

    });

File: scrape/bordercrossings.py
import scrapy

import logging


class BorderCrossingsSpider(scrapy.Spider):
    name = 'bordercrossings'

    bcs = {}

    # start_urls = ['https://en.wikipedia.org/wiki/Category:International_border_crossings']

    start_urls = ['https://en.wikipedia.org/wiki/Category:Border_crossings_of_Germany']

    #def parse(self, response):
    #    for href in response.css('#mw-subcategories ul li a::attr(href)'):
    #        country_url = response.urljoin(href.extract())
    #        yield scrapy.Request(country_url, callback=self.parse_country)

    # now on country-level

    def parse(self, response):
    	countryname = response.css('#firstHeading::text').extract()[0].replace('Category:Border crossings of ', '')
	# logging.info("Scraping %s" % countryname)
        for href in response.css('#mw-subcategories ul li a::attr(href)'):
            request = scrapy.Request(response.urljoin(href.extract()), callback=self.parse_countryborder)
            request.meta['country'] = countryname
            yield request


    #def parse_country(self, response):
    #    for href in response.css('#mw-subcategories ul li a::attr(href)'):
    #        country_border_url = response.urljoin(href.extract())
    #        yield scrapy.Request(country_border_url, callback=self.parse_countryborder)

    # now on country-border-level

    def parse_countryborder(self, response):
        for href in response.css('#mw-pages ul li a::attr(href)'):
            request = scrapy.Request(response.urljoin(href.extract()), callback=self.parse_bordercrossing)
            request.meta['country'] = response.meta['country']
            yield request

    # now on bordercrossing page

    def parse_bordercrossing(self, response):

    	bcname = response.css('#firstHeading::text').extract()[0]
	# logging.info("Scraping %s" % bcname)

        geo_url = response.css('#coordinates a.external::attr(href)').extract()[0]
        if geo_url.startswith('//'):
            geo_url = 'http:' + geo_url

        request = scrapy.Request(geo_url, callback=self.parse_geo)

        # carry all the stuff along to the final level
        request.meta['country'] = response.meta['country']
        request.meta['name'] = bcname
        request.meta['link'] = response.url

        yield request
        

    # now on geo page

    def parse_geo(self, response):

    	lat = response.css('.geo .latitude::text').extract()[0]
    	lng = response.css('.geo .longitude::text').extract()[0]

        yield {
            'name':    response.meta['name'],
            'link':    response.meta['link'],
            'country': response.meta['country'],
            'lat': lat,
            'lng': lng,
        }